src/resolvers/resolver.cr

```diff
@@ -11,33 +11,70 @@ module Shards
     def initialize(@dependency)
     end
 
-    def specs(versions)
-      specs = {} of String => Spec
-      versions.each do |version|
-        specs[version] = spec(version)
-      rescue Error
-      end
-      specs
-    end
-
     def installed_spec
       return unless installed?
 
       path = File.join(install_path, SPEC_FILENAME)
-      return Spec.from_file(path) if File.exists?(path)
+      unless File.exists?(path)
+        raise Error.new("Missing #{SPEC_FILENAME.inspect} for #{dependency.name.inspect}")
+      end
 
-      raise Error.new("Missing #{SPEC_FILENAME.inspect} for #{dependency.name.inspect}")
+      spec = Spec.from_file(path)
+      spec.version = File.read(version_path)
+      spec
     end
 
     def installed?
       File.exists?(install_path)
     end
 
-    abstract def read_spec(version = nil)
-    abstract def spec(version)
+    def versions_for(dependency : Dependency) : Array(String)
+      if ref = dependency.refs
+        versions_for_ref(ref)
+      else
+        releases = available_releases
+        if releases.empty?
+          versions_for_ref(nil)
+        elsif version_req = dependency.version?
+          Versions.resolve(releases, version_req)
+        else
+          releases
+        end
+      end
+    end
+
+    private def versions_for_ref(ref : String?) : Array(String)
+      if version = latest_version_for_ref(ref)
+        [version]
+      else
+        [] of String
+      end
+    end
+
+    abstract def available_releases : Array(String)
+    abstract def latest_version_for_ref(ref : String?) : String?
+
+    def spec(version : String) : Spec
+      spec = Spec.from_yaml(read_spec(version))
+      spec.resolver = self
+      spec.version = version
+      spec
+    end
+
+    private abstract def read_spec(version : String)
```

Comment:
Is there a reason for making this this private? It's useful as external API, for example for shardbox which stores the entire spec file in the database.

Reply:
The only reason is that `Resolver#spec` would also set the expected version in the returned `Spec` to have precise information about where that spec came from.

Are you using this API in shardbox? I can make it public, but can't you just use the `spec` method instead?

Reply:
Shardbox uses both. `spec` to get the parsed data but also `read_spec` to store the raw content of `shard.yml`. Parsing with `Spec.from_yaml` ignores unknown properties, thus loses information that might be useful to have.

Reply:
Ok, I'll make it public